src/math/exp_fcn.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * Arguments	: int64_t a
+ *		  int64_t b
+ * Return Type	: int64_t
+ */
+static inline int64_t lomul_s64_sr_sat_near(int64_t a, int64_t b)
+{
+	uint64_t u64_rhi;
+	uint64_t u64_rlo;
+	bool roundup;
+
+	mul_s64(a, b, &u64_rhi, &u64_rlo);
+	roundup = (u64_rlo & BIT_MASK_LOW_Q27P5);
+	u64_rlo = (u64_rhi << 36 | u64_rlo >> 28) + roundup;
+	return u64_rlo;
+}
+
+/*
+ * Arguments	: int64_t in_0
+ *		  int64_t in_1
+ *		  uint64_t *ptroutbitshi
+ *		  uint64_t *ptroutbitslo
+ * Multiplication of two signed int64 numbers
+ * Return Type	: void
+ */
+static inline void mul_s64(int64_t in_0, int64_t in_1, uint64_t *ptroutbitshi,
+				uint64_t *ptroutbitslo)
+{
+	uint64_t absin0 = ABS(in_0);
+	uint64_t absin1 = ABS(in_1);
+	uint64_t in0hi = absin0 >> 32;
+	uint64_t in0lo = absin0 & UINT32_MAX;
+	uint64_t in1hi = absin1 >> 32;
+	uint64_t prodhilo;
+	uint64_t prodlohi;
+
+	absin0 = absin1 & UINT32_MAX;
+	// multiplication
+	prodhilo = in0hi * absin0;
+	prodlohi = in0lo * in1hi;
+	absin0 *= in0lo;
+	absin1 = 0;
+	in0lo = absin0 + (prodlohi << 32);
+	absin1 = in0lo < absin0 ? 1 : 0;
+	absin0 = in0lo;
+	//shift and add
+	in0lo += prodhilo << 32;
+	if (in0lo < absin0)
+		absin1++;
+
+	absin0 = absin1 + in0hi * in1hi + (prodlohi >> 32) + (prodhilo >> 32);
+	// 2's compliment
+	if (in_0 != 0 && in_1 != 0 && (in_0 > 0) != (in_1 > 0)) {
+		// bit inversion
+		absin0 = ~absin0;
+		in0lo = ~in0lo;
+		// add to low byte
+		in0lo++;
+		if (!in0lo)
+			absin0++;
+	}
+	// pointer- output high and low bytes
+	*ptroutbitshi = absin0;
+	*ptroutbitslo = in0lo;
+}
+
+/* f(x) = a^x, x is variable and a is base
+ *
+ * Arguments	: int32_t x(Q4.28)
+ * input range -5 to 5
+ *
+ * Return Type	: int32_t ts(Q9.23)
+ * output range 0.0067465305 to 148.4131488800
+ *+------------------+-----------------+--------+--------+
+ *| x		     | ts (returntype) |   x    |  ts    |
+ *+----+-----+-------+----+----+-------+--------+--------+
+ *|WLen| FLen|Signbit|WLen|FLen|Signbit| Qformat| Qformat|
+ *+----+-----+-------+----+----+-------+--------+--------+
+ *| 32 | 28  |  1    | 32 | 23 |   0   | 4.28   | 9.23   |
+ *+------------------+-----------------+--------+--------+
+ */
+int32_t exp_int32(int32_t x)
+{
+	int64_t dividend = NUM_Q14P18;
+	int32_t ts =  TERMS_Q23P9; /* Q23.9 */
+	int64_t qt;
+	uint64_t ou0Hi;
+	uint64_t ou0Lo;
+	int32_t b_n;
+	// pre-computation of 1st & 2nd terms
+	dividend = (x + LSHIFT_BITS) >> 14; //x in Q50.14
+
+	mul_s64(dividend, BIT_MASK_Q62P2, &ou0Hi, &ou0Lo);
+	qt = (ou0Hi << 46) | (ou0Lo >> 18);/* Q6.26 */
+	ts += (int32_t)((qt >> 35) +
+			((qt & QUOTIENT_SCALE) >> 18));
+	dividend = lomul_s64_sr_sat_near(dividend, (int64_t)x);
+	for (b_n = 0; b_n < ARRAY_SIZE(iv); b_n++) {
+		mul_s64(dividend, iv[b_n], &ou0Hi, &ou0Lo);
+		qt = (int64_t)(ou0Hi << 45) | (ou0Lo >> 19);
+
+		// sum of the remaining terms
+		ts += (int32_t)((qt >> 35) +
+				((qt & QUOTIENT_SCALE) ? 1 : 0));
+		dividend = lomul_s64_sr_sat_near(dividend, (int64_t)x);
```

Comment:
What is the purpose of this cast?

Reply:
You're adding a cast that seems unnecessary so it's your duty to add a comment explaining what difference it makes (I'm guessing: none). Casting is a bad practice so casts must be either obvious or explained.

This is your code, not mine; it's your job to explain it not mine.

> some subtle conversion bug
> Would you kindly elaborate on the situation at hand?

Are you saying C integer conversion rules are easy? Everyone thinks they're incredibly complex and bug-prone.

Reply:
Add a comment explaining why this cast is required or remove it.